### PR TITLE
HTTP cache: implement GitHub Actions cache endpoints

### PR DESCRIPTION
**Pull request by edigaryev**
How to run locally for testing purposes:

1. Generate a JWT token, see `TestGHA`
2. Create a dummy `Dockerfile`, e.g. with `FROM debian:latest` contents
3. `docker buildx build -t whatever --cache-to type=gha,url=http://<your IP reachable from Docker VM>:12321/,token=<token generated in step 1>,scope=image .`

Resolves https://github.com/cirruslabs/cirrus-ci-agent/issues/344.

### CODE REVIEW

File: internal/http_cache/ghacache/ghacache.go [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+	if resp.StatusCode != http.StatusCreated {
+		log.Printf("GHA cache failed to upload the uploadable with ID %d: got HTTP %d\n",
+			id, resp.StatusCode)
+		writer.WriteHeader(http.StatusInternalServerError)
+
+		return
+	}
+
+	cache.uploadables.Delete(id)
```
Comment (fkorotkov) [inline internal/http_cache/ghacache/ghacache.go:205]: Should we `defer` it right after getting the `id`? Not sure if we need it in case of an error

**Review by fkorotkov — COMMENTED** [review]
Just couple of questions

File: internal/http_cache/http_cache.go [hunk 2]
```diff
@@ -43,7 +45,10 @@ func Start(taskIdentification *api.TaskIdentification) string {
 		Timeout: 10 * time.Minute,
 	}
 
-	http.HandleFunc("/", handler)
+	mux := chi.NewMux()
```
[the remark this answers is not in the document]
Comment (edigaryev) [reply internal/http_cache/http_cache.go]: Done in https://github.com/cirruslabs/cirrus-ci-agent/pull/349/commits/248bbf6b41501f3e8de0938f461b0fda83314e6e.

File: internal/http_cache/ghacache/ghacache.go (hunk 1 above)
Comment (edigaryev) [reply internal/http_cache/ghacache/ghacache.go]: Good catch, fixed in https://github.com/cirruslabs/cirrus-ci-agent/pull/349/commits/6fc098f715cda97c96789a922f04e0c46211a4aa.

File: internal/http_cache/ghacache/ghacache.go [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+func (cache *GHACache) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	cache.mux.ServeHTTP(writer, request)
+}
+
+func (cache *GHACache) get(writer http.ResponseWriter, request *http.Request) {
+	keys := strings.Split(request.URL.Query().Get("keys"), ",")
+	version := request.URL.Query().Get("version")
+
+	// The first key is used for exact matching which we support
+	httpCacheURL := cache.httpCacheURL(keys[0], version)
```
[the remark this answers is not in the document]
Comment (edigaryev) [reply internal/http_cache/ghacache/ghacache.go:56]: Added a check in https://github.com/cirruslabs/cirrus-ci-agent/pull/349/commits/4c4ca43ec4e5e5afaaa0df0a4019220cbd3360bc.

**Review by fkorotkov — APPROVED** [review]

**Pull request merged**